### PR TITLE
Add dataset and XPath query documentation

### PR DESCRIPTION
**Pull request by lognaturel**
Things I hesitated on:
- Should the datasets content go in the logic doc? That would probably be most consistent. I'd personally prefer to move towards smaller pages.
- Should I show an example of an XML dataset? Seems rare and advanced enough that it's not necessary.

### CODE REVIEW

File: src/form-logic.rst [hunk 1]
```diff
@@ -70,14 +69,84 @@ as well as any column that accepts an :ref:`expression <expressions>`.
...
+
+XPath paths can also be relative. For example, let's say there's a ``relevance`` expression for ``q1`` in the example above and that this expression refers to a question with name ``eligible`` in the ``my_group`` group. We could refer to it using an absolute expression: ``/data/my_group/eligible``. We could also write a relative expression: ``../../eligible``. The ``..`` references the parent of the current node. Since our expression is on the node ``/data/my_group/inner_group/q1``, the first ``..`` references the group ``/data/my_group/inner_group``. We want to access a node in the ``my_group`` group so we need to go up one more level with another ``..``. The XPath path ``../../`` when used in an expression for the ``/data/my_group/inner_group/q1`` node references ``/data/my_group``. We can now go into ``my_group`` and reference the question with name ``eligible``: ``../../eligible``. Relative expressions are particularly useful in the context of repeats.
```
Comment (yanokwa) [inline src/form-logic.rst:93]: I found this section very hard to parse. Here's my attempt...

> XPath paths can also be relative. For example, let's say there's a ``relevance`` expression for ``q1`` in the example above and that this expression refers to a ``group_name`` question in ``outer_group``. We could refer to ``group_name`` using an absolute expression: ``/data/outer_group/group_name``, but we could also write the relative expression ``../../group_name`` on ``q1``.
>
> The relative expression says that from your current position (`q1`), go up two levels (``../../``), and then reference ``group_name``. So in this case, the rightmost ``..`` references the parent of ``q1`` (``/data/outer_group/inner_group``) and leftmost ``..`` references the parent of that ``inner_group`` node (``/data/outer_group``). Once we have the ``outer_group``, we can reference the ``group_name`` question directly. The `/` is used to separate the different levels.

Note that I don't use ``eligible`` because it sounds like a keyword like ``relevance``.

**Review by yanokwa — CHANGES_REQUESTED** [review]
Really clear documentation as always. Hoping my changes make it even clearer.

File: src/form-logic.rst [hunk 2]
```diff
@@ -190,25 +257,9 @@ Every expression is constantly re-evaluated as an enumerator progresses through
...
-  calculate, datetime_first_opened, , once(now())
-
-The following calculate will keep track of the first time the enumerator set a value for the :th:`age` question:
-
-.. csv-table:: survey
-  :header: type, name, label, calculation
-
-  integer, age, What is your age?,
-  calculate, age_timestamp, , "if(${age} = '', '', once(now()))"
+This is particularly important to remember when using functions that access state outside of the form such as :func:`random` or :func:`now`. The value they represent may change as the conditions listed above take place.
```
[the remark this answers is not in the document]
Comment (lognaturel) [reply src/form-logic.rst:260]: It's... not defined inside? Seems really clear to me but I'll try to come up with alternatives.

File: src/form-logic.rst [hunk 3]
```diff
@@ -70,14 +70,84 @@ as well as any column that accepts an :ref:`expression <expressions>`.
...
+
+.. csv-table::
+  :header: , Explanation, Example, Notes
+  
+  ., current question's value, . >= 18, Used in :ref:`constraints <constraints>`.
+  \.\., current question's parent group, position(..), Used with :func:`position` to get the iteration index.
+
+.. _xpath-paths:
+
+Advanced: XPath paths
```
[the remark this answers is not in the document]
Comment (lognaturel) [reply src/form-logic.rst:82]: I want to make it really clear that it's optional. This page is a good doc to read from top to bottom but most people probably don't need to read this section. Do you have an alternative suggestion to achieve that goal?

File: src/form-logic.rst [hunk 4]
```diff
@@ -70,14 +69,84 @@ as well as any column that accepts an :ref:`expression <expressions>`.
...
+
+When writing expressions within a repeat, it can be helpful to use the position of the repeat instance an enumerator is currently filling out. This can be done by using the  :func:`position` function. One context in which this is useful is if you want to first collect a roster of people or things and then ask additional questions about each of those. As shown in the example in the :func:`position`, you can use a first repeat for the roster and then a second repeat that references items in the first repeat based on their position.
+
+Another application of expressions that use the ``position`` function is to access a preceding repeat instance. See an example of this in :ref:`dynamic defaults in repeats <dynamic-defaults-repeats>`.
+
+It can be useful to reference all or some repeat instances from outside the repeat. XPath notation is particularly helpful for filtering repeat instances, for example to provide a summary from data collected in repeats:
+
+.. rubric:: XLSForm
```
[the remark this answers is not in the document]
Comment (lognaturel) [reply src/form-logic.rst:110]: I disagree?!

File: src/form-logic.rst [hunk 5]
```diff
@@ -70,14 +69,84 @@ as well as any column that accepts an :ref:`expression <expressions>`.
...
+
+  begin repeat, people, Person,
+  int, age, Age,
+  int, pet_count, How many pets does this person have?
+  end repeat, people, ,
+
+  int, total_pets, , sum(${people}[age < 18]/pet_count)
+  note, total_note, Total pets owned by children: ${total_pets}
+
+In the path expression ``${people}[age < 18]/pet_count``, ``${people}`` uses ``${}`` notation to refer to all of the instances of the repeat. There is then a filter applied in ``[ ]`` brackets that only includes repeat instances for which the specified age was below 18. The ``/pet_count`` final part of the path specifies that for every included repeat instance, the specified value for ``pet_count`` should be used. This results in a set of ``pet_count`` values that can be used by the :func:`sum` function or other :ref:`functions that take nodeset arguments <repeat-functions>`. This same expression could be used in many different contexts. For example, it could define the :ref:`relevance <relevants>` of a group if there's a section of questions that only need to be filled out if there are more than one child-owned pets in the community.
```
[the remark this answers is not in the document]
Comment (lognaturel) [reply src/form-logic.rst]: I reworked this and added a section on predicates.

File: src/form-datasets.rst [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  :header: name, label, lga
+
+  eziama, Eziama, aba_n
+  umuogor, Umuogor, aba_n
+  ezeke_amasiri, Ezeke amasiri, afikpo_n
+  poperi_amasiri, Poperi amasiri, afikpo_n
+
+.. _selects-from-xml:
+
+Building selects from XML files
```
[the remark this answers is not in the document]
Comment (lognaturel) [reply src/form-datasets.rst:76]: I think it's important. It's a contrast with internal datasets defined on the choices field. We will also have `xml-internal` and `csv-internal`.

File: src/form-datasets.rst [hunk 7]
```diff
@@ -34,6 +34,10 @@ External datasets can be useful when:
   * data is reused between forms. It may be easier to attach the same data file to multiple forms instead of copying the data into all the form definitions.
   * the same forms are used in different contexts. For example, the exact same form definition could be used in multiple countries with different data files listing regions, local products, etc.
 
+.. warning::
+
+  Large datasets can affect the performance of forms, especially when complex filtering is used. We find that mobile devices released in 2019 and later can generally handle lists of tens of thousands or more without noticeable slowdowns. If you experience slow performance when using a dataset in your form, please share the size of the dataset, the device you are using, and any expressions that reference the dataset including choice filters on `the community forum <https://forum.getodk.org/c/support/6>`_.
```
Comment (yanokwa) [inline src/form-datasets.rst:39]: Have them email it to support@getodk.org instead. I think when we are asking people for stuff, we should make it as easy as possible.

File: src/form-datasets.rst [hunk 8]
```diff
@@ -98,8 +98,8 @@ The ``item`` blocks are analogous to rows in the CSV representation. Each ``item
 Referencing values in datasets
 ---------------------------------
 
-:ref:`Xpath paths <xpath-paths>` can be used to reference values in internal or external datasets. These paths will start with a call on the ``instance(<instance name>)`` function to identify which dataset is being accessed. The next part of the path is generally `/root/item` because of the `XML structure used to represent datasets for selects <selects-from-xml>`. The only exception is when using custom XML files which may have arbitrary schemas if not used for selects.
+:ref:`XPath paths <xpath-paths>` can be used to reference values in internal or external datasets. These paths will start with a call on the ``instance(<instance name>)`` function to identify which dataset is being accessed. The next part of the path is generally `/root/item` because of the `XML structure used to represent datasets for selects <selects-from-xml>`. The only exception is when using custom XML files which may have arbitrary schemas if not used for selects.
```
Comment (yanokwa) [inline src/form-datasets.rst:105]: The "start with a call on" reads too technical. Why not...
>These paths will start with the ``instance(<instance name>)`` function.

**Review by yanokwa — APPROVED** [review]
I have some minor feedback that I can put in a follow on PR.
